### ISSUE TITLE
Figure1

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#4]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.cluster.hierarchy import linkage, dendrogram
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

# =========================
# 1) Data preparation
# =========================
required_cols = [
    "age","height","BW","BMI","Hb","Hct","Alb","AST","ALT","ALP","Cre",
    "Glu","TG","HDL-C","LDL-C","LH","FSH","CRP","TT"
]

# age & BMI complete only (N=5854)
X = X_raw.loc[X_raw[["age","BMI"]].notna().all(axis=1), required_cols].copy()
print("N for dendrogram (age & BMI complete):", X.shape[0])

# Convert to numeric (to handle potential non-numeric entries)
for c in required_cols:
    X[c] = pd.to_numeric(X[c], errors="coerce")

# Treat inf as NaN
X = X.replace([np.inf, -np.inf], np.nan)

# Check missing values (for verification)
na_counts = X.isna().sum()
print("\nMissing values per column (top):")
print(na_counts[na_counts > 0].sort_values(ascending=False).head(10))

# =========================
# 2) Mean imputation → standardization (consistent with K-means)
# =========================
imputer = SimpleImputer(strategy="mean")
scaler = StandardScaler()

X_imp = imputer.fit_transform(X.values)
X_std = scaler.fit_transform(X_imp)

# Final check (ensure all values are finite)
print("\nAll finite after preprocessing:", np.isfinite(X_std).all())

# =========================
# 3) Hierarchical clustering (Ward)
# =========================
Z = linkage(X_std, method="ward")  # Ward assumes Euclidean distance, so no metric specification is needed

# =========================
# 4) Dendrogram (for manuscript: no labels, no color coding)
# =========================
plt.figure(figsize=(10, 6))
dendrogram(
    Z,
    no_labels=True,
    color_threshold=None,
)
plt.xlabel("Subjects")
plt.ylabel("Euclidean distance (Ward linkage)")
plt.title("Hierarchical clustering dendrogram (exploratory visualization)")
plt.tight_layout()
plt.show()
